contracts/modules/uniswap/v3/V3SwapRouter.sol

```diff
@@ -131,7 +141,12 @@ abstract contract V3SwapRouter is RouterImmutables, Permit2Payments, IUniswapV3S
         (int256 amount0Delta, int256 amount1Delta, bool zeroForOne) =
             _swap(-amountOut.toInt256(), recipient, path, payer, false);
 
-        uint256 amountOutReceived = zeroForOne ? uint256(-amount1Delta) : uint256(-amount0Delta);
+        uint256 amountOutReceived;
+        assembly {
+            // amountOutReceived = uint256(-(zeroForOne ? amount1Delta : amount0Delta))
+            // no need to check for underflow
+            amountOutReceived := sub(0, xor(amount0Delta, mul(xor(amount1Delta, amount0Delta), zeroForOne)))
```

Comment:
could these ternaries in this file get abstracted into the Library with all the xors??

Reply:
Actually yes. I was afraid the optimizer wouldn't inline properly. But it seems the gas level remains the same as long as we uncheck the unary `-` on `int256`
```Solidity
unchecked {
    uint256 amountOutReceived = uint256(-zeroForOne.ternary(amount1Delta, amount0Delta));
    if (amountOutReceived != amountOut) revert V3InvalidAmountOut();
}
```

For
```Solidity
hasMultiplePools ? address(this) : recipient
```
using `hasMultiplePools.ternary(address(this), recipient)` in place also saves gas due to less stack shuffling.

For `zeroForOne := xor(isExactIn, lt(tokenOut, tokenIn))` though, using
```Solidity
zeroForOne = isExactIn.ternary(tokenIn < tokenOut, tokenOut < tokenIn);
```
wouldn't be cheaper. And writing `zeroForOne = isExactIn ^ (tokenOut < tokenIn)` isn't allowed since "Operator ^ not compatible with types bool and bool.".

Writing
```Solidity
sqrtPriceLimitX96 = zeroForOne.ternary(MIN_SQRT_RATIO + 1, MAX_SQRT_RATIO - 1);
```
instead of `xor` and literal hexes in assembly also wouldn't be cheaper. We could define the literal hexes as contract level constants though if that improves readability.

Reply:
cool thanks for the explanation. We definitely strive to strike a balance between readability and gas savings (or else we could just write low-level code for everything :)  

Even with our steepest gas savings tests, these ternaries that cannot be abstracted away would only save < 100 extra gas.  So I'd personally vote to just omit them. Can always get a third opinion..

Also some context, our v2 UniversalRouter has already been audited so we cannot include this, these gas savings would go into V3 of the router.

Reply:
We can always abstract them away. It will just costs a few more opcodes. I'm okay with replacing `sqrtPriceLimitX96` done in assembly with literal hexes with `zeroForOne.ternary(MIN_SQRT_RATIO + 1, MAX_SQRT_RATIO - 1)`. But frankly speaking, is
```Solidity
zeroForOne = isExactIn ? tokenIn < tokenOut : tokenOut < tokenIn;
```
more understandable and cleaner than
```Solidity
assembly {
    zeroForOne := xor(isExactIn, lt(tokenOut, tokenIn))
}
```
with derivations?